### PR TITLE
adds a visible line when using iv drips / bloodpacks

### PR DESCRIPTION
**Pull request by realkhad**
# About the pull request

https://forum.cm-ss13.com/t/giving-blood-shows-visible-line/2610
adds a line between the person receiving blood and where the blood is coming from, similar to the phone cable

# Explain why it's good for the game

soul :]

# Testing Photographs and Procedure

https://github.com/cmss13-devs/cmss13/assets/54692343/e023add2-7172-4641-a64e-e60a313a9e1a

# Changelog
:cl:Khadd
add: added a iv tube between the user and the bloodpack / iv drip
imageadd: sprites for the iv tube
/:cl:

### CODE REVIEW

**Comment by cm13live** [comment]
This pull request has been mentioned on **CM-SS13**. Please provide feedback here:

https://forum.cm-ss13.com/t/giving-blood-shows-visible-line/2610/2

File: code/game/machinery/iv_drip.dm [hunk 1]
```diff
@@ -35,6 +36,26 @@
...
+		if(beaker)
+			beam_icon = (attached.blood_volume == 0) ? "iv_tube" : "iv_tube_blood"
+		if(current_beam && current_beam.icon_state != beam_icon || ignore)
+			delete_beam()
+			current_beam = src.beam(attached, beam_icon)
+
+/obj/structure/machinery/iv_drip/proc/delete_beam()
+	if(current_beam)
+		qdel(current_beam)
+		current_beam = null
```
Comment (harryob) [inline code/game/machinery/iv_drip.dm:57]:
```suggestion
		QDEL_NULL(current_beam)
```

File: code/game/objects/items/reagent_containers/blood_pack.dm [hunk 2]
```diff
@@ -32,6 +34,24 @@
...
+	else
+		beam_icon = (connected_to.blood_volume == 0) ? "iv_tube" : "iv_tube_blood"
+		if(current_beam && current_beam.icon_state != beam_icon || ignore)
+			delete_beam()
+			current_beam = connected_from.beam(connected_to, beam_icon)
+
+/obj/item/reagent_container/blood/proc/delete_beam()
+	if(current_beam)
+		qdel(current_beam)
+		current_beam = null
```
Comment (harryob) [inline code/game/objects/items/reagent_containers/blood_pack.dm:53]:
```suggestion
		QDEL_NULL(current_beam)
```

File: code/game/machinery/iv_drip.dm [hunk 3]
```diff
@@ -35,6 +36,26 @@
...
+			beam_icon = (beaker.reagents.total_volume == 0) ? "iv_tube" : "iv_tube_blood"
+		if(current_beam && current_beam.icon_state != beam_icon || ignore)
+			delete_beam()
+			current_beam = attached.beam(src, beam_icon)
+	else
+		if(beaker)
+			beam_icon = (attached.blood_volume == 0) ? "iv_tube" : "iv_tube_blood"
+		if(current_beam && current_beam.icon_state != beam_icon || ignore)
+			delete_beam()
+			current_beam = src.beam(attached, beam_icon)
```
Comment (harryob) [inline code/game/machinery/iv_drip.dm:52]:
```suggestion
			current_beam = beam(attached, beam_icon)
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/game/machinery/iv_drip.dm [hunk 4]
```diff
@@ -8,6 +8,7 @@
 	var/mob/living/carbon/attached = null
 	var/mode = 1 // 1 is injecting, 0 is taking blood.
 	var/obj/item/reagent_container/beaker = null
+	var/datum/beam/current_beam = null
```
Comment (harryob) [inline code/game/machinery/iv_drip.dm:11]:
```suggestion
	var/datum/beam/current_beam
```

**Review by harryob — CHANGES_REQUESTED** [review]
nothing further from me, still needs @cmss13-devs/spriters

**Review by morrowwolf — CHANGES_REQUESTED** [review]
This is really cool! Great addition.

*Sadly*, the red animated beam can look a little... off when standing above someone. I think it'd be best to just go with the basic black beam.

**Review by sleepynecrons — CHANGES_REQUESTED** [review]
just use iv_beam for both the connection and when it's active. there's no need to show the blood moving and it just looks like something out of cruelty squad. the iv_beam itself also should just use colors from the iv pole
image
something like on the right

**Review by sleepynecrons — APPROVED** [review]
sprite approved

**Comment by chapterob** [comment]
> sprite approved

My nigga brother in Christ I had to make a new account to tell you that new sprite sucks

**Comment by realkhad** [comment]
> just use iv_beam for both the connection and when it's active. there's no need to show the blood moving and it just looks like something out of cruelty squad. the iv_beam itself also should just use colors from the iv pole image something like on the right

the purpose for the animated blood sprite was to show which direction the reagents were going and if there were even any reagents being exchanged (if not it would just turn into the empty iv tube sprite)

File: code/game/machinery/iv_drip.dm [hunk 5]
```diff
@@ -4,10 +4,12 @@
 	anchored = FALSE
 	density = FALSE
 	drag_delay = 1
+	base_pixel_x = 5
```
Comment (Drulikar) [inline code/game/machinery/iv_drip.dm:7]:
```suggestion
	base_pixel_x = 15
	base_pixel_y = -2
```
image

File: code/game/machinery/iv_drip.dm [hunk 6]
```diff
@@ -57,6 +83,7 @@
 			"You attach \the [src] to \the [over_object].")
```
Comment (Drulikar) [inline code/game/machinery/iv_drip.dm:83]: Its out of scope for this PR, but it is kinda weird that the IV drip doesn't have a skill check like the blood bag does.

**Review by Drulikar — CHANGES_REQUESTED** [review]
Functionally works, just a couple things to tweak.

File: code/game/machinery/iv_drip.dm (hunk 5 above)
Comment (Drulikar) [reply code/game/machinery/iv_drip.dm:7]: Unfortunately its still kind funky that it'd be so stiff at that point to be straight and then we have a tube that goes too the target.

**Comment by sleepynecrons** [comment]
> > just use iv_beam for both the connection and when it's active. there's no need to show the blood moving and it just looks like something out of cruelty squad. the iv_beam itself also should just use colors from the iv pole image something like on the right
> 
> the purpose for the animated blood sprite was to show which direction the reagents were going and if there were even any reagents being exchanged (if not it would just turn into the empty iv tube sprite)

i know what the purpose was, it just doesn't look good in-game and you already have a HUD icon telling you if you're having blood transfused

**Review by Drulikar — APPROVED** [review]

**Pull request merged**